enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,267 @@
+---
+title: Safe OpenShift Upgrades with Operators 
+authors:
+  - "@awgreene"
+reviewers:
+  - TBD
+  - "@kevinrizza"
+approvers:
+  - "@ecordell"
+  - "@spadgett"
+creation-date: 2021-01-11
+last-updated: 2021-01-19
+status: provisional
+see-also:
+  -  N/A
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Safe OpenShift Upgrades with Operators
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Definitions
+
+**OLM Managed Operators**: Operators that are installed, managed, and upgraded by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) project.
+
+**Minor OpenShift Upgrade**: When the OpenShift Cluster is upgraded to the next minor version, for example the upgrade from version 4.6 to 4.7. See [Semantic Versioning](https://semver.org/) for more details.
+
+## Summary
+
+OpenShift Cluster Admins introduce new services to their clusters in the form of OLM Managed Operators. When performing a cluster upgrade, there is no guarantee that these operators will continue to run on the upgraded cluster.
+
+OLM must make an effort to prevent cluster upgrades when it is possible to determine that one or more installed operators will not run on the next version of OpenShift.
+
+## Motivation
+
+The primary purpose of this enhancement is to define how OpenShift can protect existing services introduced by OLM Managed Operators when performing a Minor OpenShift upgrade.
+
+### Goals
+
+- Prevent Minor OpenShift Upgrades when OLM detects that an installed operator will not run on the next Minor OpenShift version.
```

Comment:
```suggestion
- Prevent Minor OpenShift Upgrades when OLM detects that an installed operator may not run on the next Minor OpenShift version.
```

Reply:
@bparees I believe that the wording is correct in this case, OLM will prevents upgrades when it can guarantee that one or more operator will not run on the next version of operator.because of the mentioned annotation.

Reply:
perhaps i'm being too cute here, but just because an operator says "I don't want to be run on 4.7" doesn't mean it can't actually run on 4.7.  That was the intent of my suggested change.

I was also trying to cover the case where the operator doesn't tell us anything.  Perhaps that needs to be added as a second goal ("Inform the admin when one or more operators do not explicitly declare support for the next OCP version so that the admin can decide whether or not to upgrade")

Reply:
how about these two goals:

1) prevent minor version ocp upgrades when an installed operator declares that it does not support the next ocp minor version
2) warn the admin when an installed operator does not explicitly declare support for the next ocp minor version

?

Reply:
I accepted the suggested changes, please mark the conversation as resolved if applicable.